source/llassetgen-rendering/main.cpp

```diff
@@ -0,0 +1,383 @@
+        m_vao->enable(0);
+
+        m_program->setUniform("glyphs", m_samplerIndex);
+        m_program->setUniform("fontColor", m_fontColor);
+    }
+
+    virtual void deinitializeGL() override {
+        m_texture.reset(nullptr);
+
+        m_cornerBuffer.reset(nullptr);
+        m_program.reset(nullptr);
+        m_vertexShaderSource.reset(nullptr);
+        m_vertexShaderTemplate.reset(nullptr);
+        m_vertexShader.reset(nullptr);
+        m_fragmentShaderSource.reset(nullptr);
+        m_fragmentShaderTemplate.reset(nullptr);
+        m_fragmentShader.reset(nullptr);
+        m_vao.reset(nullptr);
+    }
+
+    virtual void resizeGL(QResizeEvent *event) override {
+        glViewport(0, 0, event->size().width(), event->size().height());
+    }
+
+    virtual void paintGL() override {
+        glClearColor(m_backgroundColor.r, m_backgroundColor.g, m_backgroundColor.b, m_backgroundColor.a);
+        glClear(GL_COLOR_BUFFER_BIT | GL_DEPTH_BUFFER_BIT);
+
+        glEnable(GL_BLEND);
+        glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);
+
+        if (m_texture) {
+            m_texture->bindActive(m_samplerIndex);
+        }
+
+        m_program->use();
+        m_program->setUniform("fontColor", m_fontColor);
+        m_vao->drawArrays(GL_TRIANGLE_STRIP, 0, 4);
+        m_program->release();
+
+        if (m_texture) {
+            m_texture->unbindActive(m_samplerIndex);
+        }
+
+        glDisable(GL_BLEND);
+    }
+
+    virtual void keyPressEvent(QKeyEvent *event) override {
+        makeCurrent();
+
+        switch (event->key()) {
+            case Qt::Key_F5:
+                m_vertexShaderSource->reload();
+                m_fragmentShaderSource->reload();
+                updateGL();
+                break;
+            case Qt::Key_Escape:
+                qApp->quit();
+                break;
+            default:
+                break;
+        }
+        doneCurrent();
+    }
+
+   public slots:
+    virtual void backgroundColorRChanged(QString value) override {
+        int red = value.toInt();
+        m_backgroundColor.r = red / 255.f;
+        paint();
+    }
+
+    virtual void backgroundColorGChanged(QString value) override {
+        int green = value.toInt();
+        m_backgroundColor.g = green / 255.f;
+        paint();
+    }
+
+    virtual void backgroundColorBChanged(QString value) override {
+        int blue = value.toInt();
+        m_backgroundColor.b = blue / 255.f;
+        paint();
+    }
+
+    virtual void fontColorRChanged(QString value) override {
+        int red = value.toInt();
+        m_fontColor.r = red / 255.f;
+        paint();
+    }
+
+    virtual void fontColorGChanged(QString value) override {
+        int green = value.toInt();
+        m_fontColor.g = green / 255.f;
+        paint();
+    }
+
+    virtual void fontColorBChanged(QString value) override {
+        int blue = value.toInt();
+        m_fontColor.b = blue / 255.f;
+        paint();
+    }
```

Comment:
Is there a way to deduplicate these methods? As it stands, it is basically the same code copy pasted 6 times. Maybe something like:
```cpp
void applyColorChange(float& color, QString value) {
    color = value.toInt() / 255.f;
    paint();
}

virtual void fontColorBChanged(QString value) override {
    applyColorChange(m_fontColor.b, value);
}
```

Reply:
yep, good idea, thanks!